Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32978 Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types**

Fixes #32946

Differential Revision: [D19726013](https://our.internmc.facebook.com/intern/diff/D19726013)

pbelevich committed:
Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types

[ghstack-poisoned]

yf225 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -55,7 +55,9 @@ Tensor isclose(const Tensor& self, const Tensor& other, double rtol, double atol
   auto actual_error = (self - other).abs();
-  auto max_error = atol + rtol * other.abs();
+  // The original formula `atol + rtol * other.abs()` works incorrectly when
+  // other has integral dtype and other = min_value and abs(min_value) is negative`.
```
nit: it might be even easier to read if we write it as:
```suggestion
  // `other` has integral dtype and `other == min_value` and `abs(min_value)` is negative`.
```

yf225 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -55,7 +55,9 @@ Tensor isclose(const Tensor& self, const Tensor& other, double rtol, double atol
   auto actual_error = (self - other).abs();
-  auto max_error = atol + rtol * other.abs();
+  // The original formula `atol + rtol * other.abs()` works incorrectly when
+  // other has integral dtype and other = min_value and abs(min_value) is negative`.
```
Also curious why would `abs(min_value)` even be negative?

pbelevich replied:
[If std::abs is called with an argument of type X such that std::is_unsigned<X>::value is true and X cannot be converted to int by integral promotion, the program is ill-formed.
](https://en.cppreference.com/w/cpp/numeric/math/abs)

There is no positive value corresponding to std::numeric_limits<T>::lowest() for all signed types:

`std::abs(std::numeric_limits<int64_t>::lowest()) == std::numeric_limits<int64_t>::lowest()`

pbelevich committed:
Update on "Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types"


Fixes #32946

[ghstack-poisoned]

yf225 commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -55,7 +55,10 @@ Tensor isclose(const Tensor& self, const Tensor& other, double rtol, double atol
   auto actual_error = (self - other).abs();
-  auto max_error = atol + rtol * other.abs();
+  // The original formula `atol + rtol * other.abs()` works incorrectly when
+  // `other` has integral dtype and `other == min_value` and `abs(min_value)` is negative`.
```
sorry for the nit: we can probably remove the backtick at the end of this line :/

pbelevich replied:
removed

pbelevich committed:
Update on "Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types"


Fixes #32946

Differential Revision: [D19726013](https://our.internmc.facebook.com/intern/diff/D19726013)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks so much for the fix!

pbelevich committed:
Update on "Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types"


Fixes #32946

Differential Revision: [D19726013](https://our.internmc.facebook.com/intern/diff/D19726013)

[ghstack-poisoned]

pbelevich committed:
Update on "Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types"


Fixes #32946

Differential Revision: [D19726013](https://our.internmc.facebook.com/intern/diff/D19726013)

[ghstack-poisoned]

pbelevich committed:
Update on "Fix torch::allclose to handle std::numeric_limits<T>::lowest() for integral types"


Fixes #32946

Differential Revision: [D19726013](https://our.internmc.facebook.com/intern/diff/D19726013)

[ghstack-poisoned]
